ruby_programming/automated_testing/rspec_part_one_basics.md

```diff
@@ -0,0 +1,822 @@
+  todo_list = TodoList.new("My List", ["task 1", "task 2"])
+
+  expect(todo_list.name).to eq("My List (2 tasks)")
+end
+~~~
+
+Run all the tests in the terminal once again with `$ rspec` to see all of our todo list name method tests now passing.
+
+#### The Be Matcher
+
+We'll often need to test predicate methods in our classes that return either true or false. We could use the `eq` matcher to test these methods.
+
+But RSpec has a more expressive `be` matcher which is better suited to the job and will make the tests much nicer to read as you can expect the result to *be true* or *be false* rather than equal to true or equal to false which doesn't quite read as well.
+
+To demonstrate the `be` matcher, we are going write a test for another one of the `TodoList` methods. Specifically the `completed?` method.
+
+First we need an example group where we can group the tests for the `completed?` method together.
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+
+RSpec.describe TodoList do
+
+  # omitted for brevity
+
+  describe "completed?" do
+
+  end
+end
+~~~
+
+Next we will write the individual test cases for the `completed?` method. There are only two cases that we care about.
+
+What does the method return when the list is completed and what does it return when the list is not completed:
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+describe "completed?" do
+  it "returns true when all tasks are completed" do
+  end
+
+  it "returns false when all tasks are not completed" do
+  end
+end
+~~~
+
+Now lets see how to use that `be` matcher we have been hyping up by writing the expectation for the first test case:
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+it "returns true when all tasks are completed" do
+  todo_list = TodoList.new("My List", [])
+
+  expect(todo_list.completed?).to be(true)
+end
+~~~
+
+And to finish off the tests for the `completed?` method, we will write the test for when the todo list is not completed:
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+it "returns false when all tasks are not completed" do
+  todo_list = TodoList.new("My List", ["Incomplete task"])
+
+  expect(todo_list.completed?).to be(false)
+end
+~~~
+
+Finally run all the tests once again with `$ rspec` and everything should still be passing.
+
+#### The Include Matcher
+
+The include matcher functions much like the  `include?` method you've already been using with Ruby collections. It passes the expectation if the actual value *includes* the expected value.
+
+To see how the include matcher works, we will write some tests for yet another one of the todo lists methods. This time the `add_task` method.
+
+Write the following test and make sure it works by running the tests with the   `$ rspec` command in the terminal.
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+
+RSpec.describe TodoList do
+
+  # omitted for brevity
+
+  describe "add_task" do
+	  it "adds a task to the list" do
```

Comment:
This line does not visually render with the tab/spaces aligned correctly. It copied correctly to VS Code, but I think in some places it is a tab & other places it is two spaces.

Reply:
Nice spot with these, I haven't a clue how I managed to do that 😆 they all render ok for me in vs code. Could be some config I have going on.